core/src/main/scala/org/apache/spark/api/python/PythonRunner.scala

```diff
@@ -106,6 +106,13 @@ private[spark] abstract class BasePythonRunner[IN, OUT](
     val startTime = System.currentTimeMillis
     val env = SparkEnv.get
     val localdir = env.blockManager.diskBlockManager.localDirs.map(f => f.getPath()).mkString(",")
+    // if OMP_NUM_THREADS is not explicitly set, override it with the number of cores
+    if (conf.getOption("spark.executorEnv.OMP_NUM_THREADS").isEmpty) {
+      // SPARK-28843: limit the OpenMP thread pool to the number of cores assigned to this executor
+      // this avoids high memory consumption with pandas/numpy because of a large OpenMP thread pool
+      // see https://github.com/numpy/numpy/issues/10455
+      conf.getOption("spark.executor.cores").foreach(envVars.put("OMP_NUM_THREADS", _))
```

Comment:
Seems `spark.executor.cores` can mean different things, for instance,

> spark.executor.cores | 1 in YARN mode, all the available cores on the worker in standalone

Can we investigate such cases?

Reply:
Yes, so the problem here is that the number is somewhat a bit arbitrary. It should be yes 1 or up to users control.

Am I understanding correctly that, say in 16 cores machine,

spark.executor.cores=1-> all cores, OMP_NUM_THREADS=1 -> one process with 1 thread
spark.executor.cores=2 -> two cores, OMP_NUM_THREADS=2 -> one process with 2 threads
spark.executor.cores=3 -> three cores, OMP_NUM_THREADS=3 -> one process with 3 threads
...
spark.executor.cores=16-> all cores, OMP_NUM_THREADS=16 -> one process with 16 threads

is an expected behaviour?


UDF or RDD APIs cannot be optimized because we don't know what's users are going to do in in this box. And here we're putting an assumption on that.

Reply:
Also, to be clear, this is already configurable. If we're going to put some default number for all general PySpark computation, please pease let's clarify what it means in PR description and upgrade migration guide.